neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/core/Id.java

```diff
@@ -1,41 +1,85 @@
 /*
- * Copyright (c) 2013-2017 Atlanmod INRIA LINA Mines Nantes.
- * All rights reserved. This program and the accompanying materials
- * are made available under the terms of the Eclipse Public License v1.0
- * which accompanies this distribution, and is available at
- * http://www.eclipse.org/legal/epl-v10.html
+ * Copyright (c) 2013-2017 Atlanmod, Inria, LS2N, and IMT Nantes.
  *
- * Contributors:
- *     Atlanmod INRIA LINA Mines Nantes - initial API and implementation
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License v2.0 which accompanies
+ * this distribution, and is available at https://www.eclipse.org/legal/epl-2.0/
  */
 
 package fr.inria.atlanmod.neoemf.core;
 
 import java.io.Serializable;
 
 import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import javax.annotation.concurrent.Immutable;
 
 /**
  * An identifier used to identify {@link PersistentEObject}s in datastores.
  * <p>
- * This is the primary key in datastores, so, each {@code Id} must be unique at back-end level to ensure data
- * consistency. The unicity is not guaranteed before adding a new {@link PersistentEObject} to a database.
+ * This is the primary key in datastores, so, each {@code Id} must be unique at {@link
+ * fr.inria.atlanmod.neoemf.data.mapping.DataMapper} level to ensure data consistency. The unicity is not guaranteed
+ * before adding a new {@link PersistentEObject} to a database.
  */
+@Immutable
+@ParametersAreNonnullByDefault
 public interface Id extends Comparable<Id>, Serializable {
 
     /**
-     * Returns the {@link String} representation of this {@code Id}.
+     * An immutable undefined {@link Id}.
+     */
+    @Nonnull
+    Id UNDEFINED = new Id() {
+
+        /**
+         * The message of the exception thrown when accessing an unsupported method.
+         */
+        private final String msg = "Undefined identifier";
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Make this final field static too. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)